Improving stage

KieranWoodTombola opened the pull request:
Used more assets from the cdn to improve the game's visuals.
Line changes make it look like there are dramatic changes but that's mostly from importing JSON files.

Known Issues:
Text formatting fails to load on the first menu scene, but loads after.

KieranWoodTombola committed:
Feature: Added a Grid Background

- Added background of spinning tokens from Vikings

- Made changes to grid, column and token to allow tokens
to tween to the correct positions

- Made changes to grid and column to allow tokens to be placed
on the new background

KieranWoodTombola committed:
Fix: Adjusted curve of tweening tokens on grid

KieranWoodTombola committed:
Feature: Added backgrounds to game stats

- Added backgrounds to Timer and ScoreDisplay

- Removed some obsolete assets

- Repositioned elements to better fit the stage

KieranWoodTombola committed:
Feature: Added more ships to background

- Changed Sprite to the wild-static sprite to differentiate
from tokens used in the grid

- Added more Sprites to the background for post-game tweens

KieranWoodTombola committed:
Reformatting: Improving Readability

- Refactored the timer to call an event on the timer reaching 0,
rather than relying on callbacks as callbacks make garbage collection
awkward

- Refactored object loading on scenes into seperate classes. Previously,
all objects were loaded in the scene's onLoad function which made
building scene transitions awkward

- Fixed an issue where events weren't being picked up by garbage collecter

- Fixed an issue where all matched tokens would move to position 0
in their arrays

- Fixed an issue where the last tokens in rows and columns weren't
being checked

- Moved some files around

KieranWoodTombola committed:
Feature: Added bett backgrounds for scenes

- Hijacked the Vikings intro assets to improve the scene
background

KieranWoodTombola committed:
Feature: Added Lightning Effects

- Added onMatch listener to background so that Lightning
plays on a succesfful match

- Lightning plays on loop in startup menu

KieranWoodTombola committed:
Feature: Added Highscore to start scene

- Added scoreDisplay to show current highscore on the start menu

- Tidied imports across scenes for readability, removing some that
are now obsolete

KieranWoodTombola committed:
Merge branch 'main' of github.com:KieranWoodTombola/Match-Three into Improving_Stage

KieranWoodTombola committed:
Fix getter broken after merge

AndrewMollett commented on src/assets/animations/introduction/introduction.atlas:
```diff
@@ -0,0 +1,244 @@
+  size: 179, 248
+  orig: 193, 262
+  offset: 8, 6
+  index: -1
```
Is there a restriction for loading assets using the CDN for this project? If not, I think it would be a good task to learn more about how we load assets in our games through the CDN, it's functionally just a remote folder structure.

Loading assets through the CDN has a number of huge benefits, mainly the file structure for the game itself is far far smaller, making it quicker to load the core of things, as well as the bundle being smaller. It also allows for asset changes to occur, without the need for code updates (Assuming they still follow the same structure!)

Perhaps not as part of this task, but I think it would be a good task to learn

AndrewMollett commented on src/scripts/core/models/background.ts:
```diff
@@ -1,92 +1,175 @@
+
+    public stopThunder(): void {
+        this._thunderFlash.alpha = 0;
+        this._lightning.alpha = 0;
```
In pixi there is actually a strange bug with setting the alpha like this, where animations may play strangely when setting the alpha from 0 to 1. In the past we've actually set the alpha to 0.001 instead, which is functionally the same for all intents and purposes. But bypasses these bugs, obviously if there are no weird issues with your animations, don't do this. But it is worth keeping in mind

Said issue also happens if you set the 'renderable' property from false to true as well

AndrewMollett commented on src/scripts/core/models/background.ts:
```diff
@@ -1,92 +1,175 @@
+            angle: 20,
+        });
+
+        return ship
```
This function name implies you are adding the ship to the stage, but you're starting an animation before it has been added to the stage. I would argue this function should call addChild for the ship before you run the animation.

Heck it feels like the ships should be their own class which extends sprite, where the animations can be handled.

AndrewMollett commented on src/scripts/core/models/grid.ts:
```diff
@@ -8,33 +8,44 @@ gsap.registerPlugin(MotionPathPlugin);
-
-    private _availWidth: number;
+    private _background: Sprite = new Sprite(Assets.get('gridBackground'));
+    private _columnContainer = new Container();
```
Generally I wouldn't recommend 'newing' these here, this is functionally similar (but not the same) as 'newing' them in the constructor. Most importantly you don't have complete control over the order in which they are created here. Generally it's a good idea to only assign when declaring if it has a simple default, such as a number or string.

AndrewMollett replied:
Was a discussion had about this with someone else? Unsure why it has been marked as resolved without a response or a change?

KieranWoodTombola committed:
PR: Responding to changes

- Reformatted Background using a new Class: Ship

- Moved lines newing objects into Constructors

AndrewMollett commented on src/scripts/core/models/background.ts:
```diff
@@ -1,33 +1,24 @@
 
-    private _thunderFlash: Spine = new Spine(Assets.get("introduction").spineData);
-    private _lightning: Spine = new Spine(Assets.get("introduction").spineData);
+    private _stormCloudsContainer: Container = new Container();
```
Still lots of newing of containers at this point rather than in the constructor

AndrewMollett commented on src/scripts/core/models/background.ts:
```diff
@@ -1,33 +1,24 @@
     private _viewHeight: number;
 
-    private _stormCloudsContainer: Container = new Container();
+    private _ships: Ship[] | undefined;
```
Rather than setting this to sometimes be undefined, I would initialise it as an empty array here. If you have any undefined checks, you can instead check whether the array has a length greater than 0

AndrewMollett commented on src/scripts/core/models/ship.ts:
```diff
@@ -0,0 +1,36 @@
+            y: y
+        }
+
+        gsap.fromTo(this, {
```
I would put this in a public function that is triggered by its parent, it still feels strange to be triggering a tween like this before you have added it to the stage

KieranWoodTombola committed:
PR: Responding to comments

- Removed newing of Containers at property instantiation

- Changed type of class proerties in Ship

- Removed Tweens from class constructor in Ship
